src/cargo/core/global_cache_tracker.rs

```diff
@@ -0,0 +1,1754 @@
+            REGISTRY_CRATE_TABLE,
+            &base.crate_dir,
+        )?;
+        Self::update_db_for_removed(
+            conn,
+            REGISTRY_INDEX_TABLE,
+            "registry_id",
+            REGISTRY_SRC_TABLE,
+            &base.src,
+        )?;
+        Self::update_db_for_removed(conn, GIT_DB_TABLE, "git_id", GIT_CO_TABLE, &base.git_co)?;
+
+        // For registry_index and git_db, remove anything from the db that
+        // isn't on disk.
+        //
+        // This also collects paths for any child files that don't have their
+        // respective parent on disk.
+        Self::update_db_parent_for_removed_from_disk(
+            conn,
+            REGISTRY_INDEX_TABLE,
+            &base.index,
+            &[&base.crate_dir, &base.src],
+            delete_paths,
+        )?;
+        Self::update_db_parent_for_removed_from_disk(
+            conn,
+            GIT_DB_TABLE,
+            &base.git_db,
+            &[&base.git_co],
+            delete_paths,
+        )?;
+
+        // For registry_crate, registry_src, and git_checkout, add anything
+        // that is missing in the db.
+        Self::populate_untracked_crate(conn, &base.crate_dir)?;
+        Self::populate_untracked(
+            conn,
+            config,
+            REGISTRY_INDEX_TABLE,
+            "registry_id",
+            REGISTRY_SRC_TABLE,
+            &base.src,
+            sync_size,
+        )?;
+        Self::populate_untracked(
+            conn,
+            config,
+            GIT_DB_TABLE,
+            "git_id",
+            GIT_CO_TABLE,
+            &base.git_co,
+            sync_size,
+        )?;
+
+        // Update any NULL sizes if needed.
+        if sync_size {
+            Self::update_null_sizes(
+                conn,
+                config,
+                REGISTRY_INDEX_TABLE,
+                "registry_id",
+                REGISTRY_SRC_TABLE,
+                &base.src,
+            )?;
+            Self::update_null_sizes(
+                conn,
+                config,
+                GIT_DB_TABLE,
+                "git_id",
+                GIT_CO_TABLE,
+                &base.git_co,
+            )?;
+        }
+        Ok(())
+    }
+
+    /// For parent tables, add any entries that are on disk but aren't tracked in the db.
+    fn update_parent_for_missing_from_db(
+        conn: &Connection,
+        parent_table_name: &str,
+        base_path: &Path,
+    ) -> CargoResult<()> {
+        let _p = crate::util::profile::start(format!(
+            "update parent db for missing from db {parent_table_name}"
+        ));
+        trace!("checking for untracked parent to add to {parent_table_name}");
+        let names = Self::names_from(base_path)?;
+
+        let mut stmt = conn.prepare_cached(&format!(
+            "INSERT INTO {parent_table_name} (name, timestamp)
+                VALUES (?1, ?2)
+                ON CONFLICT DO NOTHING",
+        ))?;
+        let now = now();
```

Comment:
`DeferredGlobalLastUse` had a single instance of `now()`.  I'm seeing a lot here as well.  Should the global use tracker have a single member?  Maybe we even have the config do a single `now` for the process and initialize both of these with it?

Reply:
I don't think it is called often enough to be noticeable from a performance standpoint. Passing the timestamp around ends up adding more lines of code which I would prefer to avoid.

As for tracking something in `Config`, I considered that, but I'd be a bit uneasy with dealing with long-lived programs or handling before/after relationships (like different parts of the code not considering that all timestamps might be equal). I think it's worth considering, but I'm not sure this specific PR really justifies it balanced with the risks.

Reply:
I'm also a bit uneasy about the different `now`s used but I can't think of a concrete problem.

And cargo-as-a-library is already shot for long running applications because of `InternedString`.

However, this is likely not enough of a deal that we can pass on this.